From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tim Michals <tcmichals@gmail.com>
Date: Sun, 16 Aug 2026 11:13:00 -0500
Subject: [PATCH 3/3] wifi: aic: fix SDIO phase timing and chip wakeup sequence

On Allwinner sun55i/sun60i MMC host controllers (and standard SDR /
High-Speed 25MHz/50MHz SDIO), unconditionally writing to internal IOPAD
delay registers (0xF0, 0xF8, 0xF1) shifts the clock-to-data sampling phase,
causing MMC data CRC errors ("data error, sending stop command").

Additionally, the AIC8800D80 chip powers up in bus sleep mode; sending
FIFO data writes before an explicit aicwf_sdio_wakeup() sequence causes
transfers to time out.

1. Restrict IOPAD delay register modifications to MMC_TIMING_UHS_DDR50.
2. Add explicit aicwf_sdio_wakeup() in probe and firmware init routines.
3. Fallback gracefully to standard AIC8800D80 U02 revision if the pre-boot
   0x40500000 IPC read does not respond in cold BootROM state.

Signed-off-by: Tim Michals <tcmichals@gmail.com>
---
 .../aic8800_bsp/aic_chip_8800d80.c            | 17 ++++++++--
 aic8800_bsp/aicsdio.c                         | 31 +++++++++---------
 aic8800_fdrv/aicwf_sdio.c                     | 31 +++++++++---------
 3 files changed, 45 insertions(+), 34 deletions(-)

diff --git a/aic8800_bsp/aic_chip_8800d80.c b/aic8800_bsp/aic_chip_8800d80.c
index d25afbc..da292cb 100644
--- a/aic8800_bsp/aic_chip_8800d80.c
+++ b/aic8800_bsp/aic_chip_8800d80.c
@@ -18,6 +18,8 @@ static int aic8800d80_wifi_init(struct aic_sdio_dev *sdiodev, int testmode)
 {
 	const char *fw_path = aicbsp_firmware_list[aicbsp_info.cpmode].wl_fw;
 
+	aicwf_sdio_wakeup(sdiodev);
+
 	if (rwnx_plat_bin_fw_upload(sdiodev, RAM_FMAC_FW_ADDR,
 				    fw_path)) {
 		pr_err("aicbsp 8800d80 download wifi fw fail\n");
@@ -70,14 +72,21 @@ static int aic8800d80_driver_fw_init(struct aic_sdio_dev *sdiodev, u32 *btenable
 	u32 mem_addr;
 	struct dbg_mem_read_cfm rd_mem_addr_cfm;
 	u8 is_chip_id_h = 0;
+	int ret;
 
-	mem_addr = 0x40500000;
+	aicwf_sdio_wakeup(sdiodev);
 
-	if (rwnx_send_dbg_mem_read_req(sdiodev, mem_addr, &rd_mem_addr_cfm))
-		return -1;
+	mem_addr = 0x40500000;
 
-	aicbsp_info->chip_rev = (u8)((rd_mem_addr_cfm.memdata >> 16) & 0x3F);
-	is_chip_id_h = (u8)(((rd_mem_addr_cfm.memdata >> 16) & 0xC0) == 0xC0);
+	ret = rwnx_send_dbg_mem_read_req(sdiodev, mem_addr, &rd_mem_addr_cfm);
+	if (ret) {
+		pr_warn("aicbsp: 0x40500000 read failed (%d), defaulting to AIC8800D80 U02\n", ret);
+		aicbsp_info->chip_rev = CHIP_REV_U02;
+		is_chip_id_h = 0;
+	} else {
+		aicbsp_info->chip_rev = (u8)((rd_mem_addr_cfm.memdata >> 16) & 0x3F);
+		is_chip_id_h = (u8)(((rd_mem_addr_cfm.memdata >> 16) & 0xC0) == 0xC0);
+	}
 	*btenable = 1;
 	if (is_chip_id_h) {
 		AICWFDBG(LOGINFO, "IS_CHIP_ID_H\n");
diff --git a/aic8800_bsp/aicsdio.c b/aic8800_bsp/aicsdio.c
index aee4b41..96e133a 100644
--- a/aic8800_bsp/aicsdio.c
+++ b/aic8800_bsp/aicsdio.c
@@ -266,6 +266,7 @@ static int aicbsp_sdio_probe(struct sdio_func *func,
 #endif
 
 	aicbsp_platform_init(sdiodev);
+	aicwf_sdio_wakeup(sdiodev);
 
 	complete(aicbsp_probe_completion);
 
@@ -1722,29 +1723,28 @@ int aicwf_sdiov3_func_init(struct aic_sdio_dev *sdiodev)
 		sdio_release_host(sdiodev->func);
 		return ret;
 	}
-	// mod sdio phase
-	if (host->ios.timing == MMC_TIMING_UHS_DDR50)
-		val = 0x20; // 0x21; //0x1D; //0x5;
-	else
-		val = 0x00; // 0x01; //0x19; //0x1;
-	val |= SDIOCLK_FREE_RUNNING_BIT;
-	sdio_f0_writeb(sdiodev->func, val, 0xF0, &ret);
-	if (ret) {
-		sdio_err("set iopad ctrl fail %d\n", ret);
-		sdio_release_host(sdiodev->func);
-		return ret;
-	}
-	sdio_f0_writeb(sdiodev->func, 0x00, 0xF8, &ret);
-	if (ret) {
-		sdio_err("set iopad delay2 fail %d\n", ret);
-		sdio_release_host(sdiodev->func);
-		return ret;
-	}
-	sdio_f0_writeb(sdiodev->func, 0x00, 0xF1, &ret);
-	if (ret) {
-		sdio_err("set iopad delay1 fail %d\n", ret);
-		sdio_release_host(sdiodev->func);
-		return ret;
+	// mod sdio phase only in DDR50 mode; leave default for SDR / HighSpeed to avoid sunxi-mmc data errors
+	if (host->ios.timing == MMC_TIMING_UHS_DDR50) {
+		val = 0x20;
+		val |= SDIOCLK_FREE_RUNNING_BIT;
+		sdio_f0_writeb(sdiodev->func, val, 0xF0, &ret);
+		if (ret) {
+			sdio_err("set iopad ctrl fail %d\n", ret);
+			sdio_release_host(sdiodev->func);
+			return ret;
+		}
+		sdio_f0_writeb(sdiodev->func, 0x00, 0xF8, &ret);
+		if (ret) {
+			sdio_err("set iopad delay2 fail %d\n", ret);
+			sdio_release_host(sdiodev->func);
+			return ret;
+		}
+		sdio_f0_writeb(sdiodev->func, 0x00, 0xF1, &ret);
+		if (ret) {
+			sdio_err("set iopad delay1 fail %d\n", ret);
+			sdio_release_host(sdiodev->func);
+			return ret;
+		}
 	}
 	usleep_range(1000, 2000);
 	sdio_dbg("Set SDIO Clock %d MHz\n", host->ios.clock / 1000000);
diff --git a/aic8800_fdrv/aicwf_sdio.c b/aic8800_fdrv/aicwf_sdio.c
index 6bd1279..803d1bb 100644
--- a/aic8800_fdrv/aicwf_sdio.c
+++ b/aic8800_fdrv/aicwf_sdio.c
@@ -2526,28 +2526,28 @@ int aicwf_sdiov3_func_init(struct aic_sdio_dev *sdiodev)
 		sdio_release_host(sdiodev->func);
 		return ret;
 	}
-	if (host->ios.timing == MMC_TIMING_UHS_DDR50)
-		val = 0x20; // 0x21; //0x1D; //0x5;
-	else
-		val = 0x00; // 0x01; //0x19; //0x1;
-	val |= SDIOCLK_FREE_RUNNING_BIT;
-	sdio_f0_writeb(sdiodev->func, val, 0xF0, &ret);
-	if (ret) {
-		sdio_err("set iopad ctrl fail %d\n", ret);
-		sdio_release_host(sdiodev->func);
-		return ret;
-	}
-	sdio_f0_writeb(sdiodev->func, 0x0, 0xF8, &ret);
-	if (ret) {
-		sdio_err("set iopad delay2 fail %d\n", ret);
-		sdio_release_host(sdiodev->func);
-		return ret;
-	}
-	sdio_f0_writeb(sdiodev->func, 0x00, 0xF1, &ret);
-	if (ret) {
-		sdio_err("set iopad delay1 fail %d\n", ret);
-		sdio_release_host(sdiodev->func);
-		return ret;
+	// mod sdio phase only in DDR50 mode; leave default for SDR / HighSpeed to avoid sunxi-mmc data errors
+	if (host->ios.timing == MMC_TIMING_UHS_DDR50) {
+		val = 0x20;
+		val |= SDIOCLK_FREE_RUNNING_BIT;
+		sdio_f0_writeb(sdiodev->func, val, 0xF0, &ret);
+		if (ret) {
+			sdio_err("set iopad ctrl fail %d\n", ret);
+			sdio_release_host(sdiodev->func);
+			return ret;
+		}
+		sdio_f0_writeb(sdiodev->func, 0x0, 0xF8, &ret);
+		if (ret) {
+			sdio_err("set iopad delay2 fail %d\n", ret);
+			sdio_release_host(sdiodev->func);
+			return ret;
+		}
+		sdio_f0_writeb(sdiodev->func, 0x00, 0xF1, &ret);
+		if (ret) {
+			sdio_err("set iopad delay1 fail %d\n", ret);
+			sdio_release_host(sdiodev->func);
+			return ret;
+		}
 	}
 	usleep_range(1000, 1500);
 	sdio_release_host(sdiodev->func);
-- 
2.43.0
